File: sdk/luma/memory.py
from typing import Any, Dict, List, Optional

from ._http import Http


class MemoryClient:
    """NS-Mem agent memory layer — episodic, semantic, procedural and working memory."""

    def __init__(self, http: Http, namespace: str):
        self._http = http
        self._ns = f"/v1/memory/{namespace}"

    # ── ingest_event ─────────────────────────────────────────────────────────

    # Store an episodic event. Triggers the LLM consolidation pipeline when enabled.
    def ingest_event(self, text: str, *, id: Optional[str] = None,
                     entity_id: Optional[str] = None, metadata: Optional[Dict] = None,
                     source: Optional[str] = None, session_id: Optional[str] = None,
                     expires_at_ms: Optional[int] = None) -> dict:
        body: Dict = {"text": text}
        _opt(body, id=id, entity_id=entity_id, metadata=metadata,
             source=source, session_id=session_id, expires_at_ms=expires_at_ms)
        return self._http.post(f"{self._ns}/ingest_event", body)

    async def aingest_event(self, text: str, *, id: Optional[str] = None,
                            entity_id: Optional[str] = None, metadata: Optional[Dict] = None,
                            source: Optional[str] = None, session_id: Optional[str] = None,
                            expires_at_ms: Optional[int] = None) -> dict:
        body: Dict = {"text": text}
        _opt(body, id=id, entity_id=entity_id, metadata=metadata,
             source=source, session_id=session_id, expires_at_ms=expires_at_ms)
        return await self._http.apost(f"{self._ns}/ingest_event", body)

    # ── upsert_fact ──────────────────────────────────────────────────────────

    # Create or update a semantic fact. Versions the previous value and creates a Supersedes edge on overwrite.
    def upsert_fact(self, content: str, *, id: Optional[str] = None,
                    entity_id: Optional[str] = None, fact_key: Optional[str] = None,
                    metadata: Optional[Dict] = None, source: Optional[str] = None,
                    confidence: Optional[float] = None, status: Optional[str] = None) -> dict:
        body: Dict = {"content": content}
        _opt(body, id=id, entity_id=entity_id, fact_key=fact_key,
             metadata=metadata, source=source, confidence=confidence, status=status)
        return self._http.post(f"{self._ns}/upsert_fact", body)

    async def aupsert_fact(self, content: str, *, id: Optional[str] = None,
                           entity_id: Optional[str] = None, fact_key: Optional[str] = None,
                           metadata: Optional[Dict] = None, source: Optional[str] = None,
                           confidence: Optional[float] = None, status: Optional[str] = None) -> dict:
        body: Dict = {"content": content}
        _opt(body, id=id, entity_id=entity_id, fact_key=fact_key,
             metadata=metadata, source=source, confidence=confidence, status=status)
        return await self._http.apost(f"{self._ns}/upsert_fact", body)

    # ── upsert_procedure ─────────────────────────────────────────────────────

    # Register or replace a procedural DAG (nodes + typed edges + constraints).
    def upsert_procedure(self, procedure_id: str, name: str,
                         nodes: List[Dict], edges: List[Dict], *,
                         version: Optional[int] = None, status: Optional[str] = None,
                         description: Optional[str] = None, confidence: Optional[float] = None,
                         source: Optional[str] = None,
                         constraints: Optional[List[Dict]] = None) -> dict:
        body: Dict = {"procedure_id": procedure_id, "name": name, "nodes": nodes, "edges": edges}
        _opt(body, version=version, status=status, description=description,
             confidence=confidence, source=source, constraints=constraints)
        return self._http.post(f"{self._ns}/upsert_procedure", body)

    async def aupsert_procedure(self, procedure_id: str, name: str,
                                nodes: List[Dict], edges: List[Dict], *,
                                version: Optional[int] = None, status: Optional[str] = None,
                                description: Optional[str] = None,
                                confidence: Optional[float] = None,
                                source: Optional[str] = None,
                                constraints: Optional[List[Dict]] = None) -> dict:
        body: Dict = {"procedure_id": procedure_id, "name": name, "nodes": nodes, "edges": edges}
        _opt(body, version=version, status=status, description=description,
             confidence=confidence, source=source, constraints=constraints)
        return await self._http.apost(f"{self._ns}/upsert_procedure", body)

    # ── query ────────────────────────────────────────────────────────────────

    # Query memory across episodic, semantic and procedural layers. Mode is auto-detected from query content.
    def query(self, query: str, *, entity_id: Optional[str] = None,
              session_id: Optional[str] = None, procedure_id: Optional[str] = None,
              current_node_id: Optional[str] = None, context: Optional[Dict] = None,
              mode: Optional[str] = None, limit: Optional[int] = None,
              include_evidence: Optional[bool] = None, include_plan: Optional[bool] = None,
              include_diagnostics: Optional[bool] = None) -> dict:
        body: Dict = {"query": query}
        _opt(body, entity_id=entity_id, session_id=session_id,
             procedure_id=procedure_id, current_node_id=current_node_id,
             context=context, mode=mode, limit=limit,
             include_evidence=include_evidence, include_plan=include_plan,
             include_diagnostics=include_diagnostics)
        return self._http.post(f"{self._ns}/query", body)

    async def aquery(self, query: str, *, entity_id: Optional[str] = None,
                     session_id: Optional[str] = None, procedure_id: Optional[str] = None,
                     current_node_id: Optional[str] = None, context: Optional[Dict] = None,
                     mode: Optional[str] = None, limit: Optional[int] = None,
                     include_evidence: Optional[bool] = None, include_plan: Optional[bool] = None,
                     include_diagnostics: Optional[bool] = None) -> dict:
        body: Dict = {"query": query}
        _opt(body, entity_id=entity_id, session_id=session_id,
             procedure_id=procedure_id, current_node_id=current_node_id,
             context=context, mode=mode, limit=limit,
             include_evidence=include_evidence, include_plan=include_plan,
             include_diagnostics=include_diagnostics)
        return await self._http.apost(f"{self._ns}/query", body)

    # ── next_step ────────────────────────────────────────────────────────────

    # Return the next valid node in a procedure given current_node_id and evaluation context.
    def next_step(self, procedure_id: str, *, current_node_id: Optional[str] = None,
                  context: Optional[Dict] = None) -> dict:
        body: Dict = {"procedure_id": procedure_id}
        _opt(body, current_node_id=current_node_id, context=context)
        return self._http.post(f"{self._ns}/next_step", body)

    async def anext_step(self, procedure_id: str, *, current_node_id: Optional[str] = None,
                         context: Optional[Dict] = None) -> dict:
        body: Dict = {"procedure_id": procedure_id}
        _opt(body, current_node_id=current_node_id, context=context)
        return await self._http.apost(f"{self._ns}/next_step", body)

    # ── timeline ─────────────────────────────────────────────────────────────

    # Fetch the episodic timeline for an entity, ordered chronologically.
    def timeline(self, entity_id: str) -> dict:
        return self._http.get(f"{self._ns}/timeline/{entity_id}")

    async def atimeline(self, entity_id: str) -> dict:
        return await self._http.aget(f"{self._ns}/timeline/{entity_id}")

    # ── graph edges ──────────────────────────────────────────────────────────

    # Create a typed weighted edge between two memory records.
    def create_edge(self, source_id: str, target_id: str, edge_type: str, *,
                    id: Optional[str] = None, weight: Optional[float] = None,
                    metadata: Optional[Dict] = None) -> dict:
        body: Dict = {"source_id": source_id, "target_id": target_id, "edge_type": edge_type}
        _opt(body, id=id, weight=weight, metadata=metadata)
        return self._http.post(f"{self._ns}/edges", body)

    async def acreate_edge(self, source_id: str, target_id: str, edge_type: str, *,
                           id: Optional[str] = None, weight: Optional[float] = None,
                           metadata: Optional[Dict] = None) -> dict:
        body: Dict = {"source_id": source_id, "target_id": target_id, "edge_type": edge_type}
        _opt(body, id=id, weight=weight, metadata=metadata)
        return await self._http.apost(f"{self._ns}/edges", body)

    # Return all outgoing and incoming edges for a memory node.
    def node_edges(self, memory_id: str) -> dict:
        return self._http.get(f"{self._ns}/edges/{memory_id}")

    async def anode_edges(self, memory_id: str) -> dict:
        return await self._http.aget(f"{self._ns}/edges/{memory_id}")

    # Delete a memory edge by its ID.
    def delete_edge(self, edge_id: str) -> dict:
        return self._http.post(f"{self._ns}/edges/{edge_id}/delete")

    async def adelete_edge(self, edge_id: str) -> dict:
        return await self._http.apost(f"{self._ns}/edges/{edge_id}/delete")

    # Return the full version history of a belief (semantic fact) by fact_key.
    def belief_history(self, fact_key: str) -> dict:
        return self._http.get(f"{self._ns}/beliefs/{fact_key}/history")

    async def abelief_history(self, fact_key: str) -> dict:
        return await self._http.aget(f"{self._ns}/beliefs/{fact_key}/history")

    # Recompute PageRank centrality scores for all nodes in the namespace.
    def refresh_centrality(self) -> dict:
        return self._http.post(f"{self._ns}/graph/centrality")

    async def arefresh_centrality(self) -> dict:
        return await self._http.apost(f"{self._ns}/graph/centrality")


def _opt(body: Dict, **kwargs: Any) -> None:
    for k, v in kwargs.items():
        if v is not None:
            body[k] = v

File: sdk/typescript/src/memory.ts
import { HttpClient } from "./http.js";
import type {
  BeliefVersion,
  CentralityResult,
  CreateEdgeRequest,
  IngestEventOptions,
  MemoryEdge,
  MemoryQueryOptions,
  MemoryQueryResult,
  MemoryRecord,
  NextStepOptions,
  NodeEdgesResult,
  Procedure,
  ProcedureStep,
  UpsertFactOptions,
} from "./types.js";

/**
 * MemoryClient — NS-Mem agent memory layer for a single namespace.
 * Covers episodic, semantic, procedural, and working memory.
 *
 * Obtain via `client.memory(namespace)`.
 */
export class MemoryClient {
  private readonly ns: string;

  constructor(
    private readonly http: HttpClient,
    namespace: string,
  ) {
    this.ns = `/v1/memory/${encodeURIComponent(namespace)}`;
  }

  // ── Episodic events ──────────────────────────────────────────────────────

  /**
   * Store an episodic event. Triggers the LLM consolidation pipeline when
   * enabled (extracts semantic facts from the event text).
   */
  ingestEvent(
    text: string,
    opts: IngestEventOptions = {},
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { text };
    setOpt(body, "id", opts.id);
    setOpt(body, "entity_id", opts.entity_id);
    setOpt(body, "metadata", opts.metadata);
    setOpt(body, "source", opts.source);
    setOpt(body, "session_id", opts.session_id);
    setOpt(body, "expires_at_ms", opts.expires_at_ms);
    return this.http.post(`${this.ns}/ingest_event`, body);
  }

  // ── Semantic facts ───────────────────────────────────────────────────────

  /**
   * Create or update a semantic fact. Versions the previous value and creates
   * a Supersedes (or Contradicts) edge on overwrite.
   */
  upsertFact(
    content: string,
    opts: UpsertFactOptions = {},
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { content };
    setOpt(body, "id", opts.id);
    setOpt(body, "entity_id", opts.entity_id);
    setOpt(body, "fact_key", opts.fact_key);
    setOpt(body, "metadata", opts.metadata);
    setOpt(body, "source", opts.source);
    setOpt(body, "confidence", opts.confidence);
    setOpt(body, "status", opts.status);
    return this.http.post(`${this.ns}/upsert_fact`, body);
  }

  // ── Procedural memory ────────────────────────────────────────────────────

  /**
   * Register or replace a procedural DAG (nodes + typed edges + constraints).
   */
  upsertProcedure(proc: Procedure): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {
      procedure_id: proc.procedure_id,
      name: proc.name,
      nodes: proc.nodes,
      edges: proc.edges,
    };
    setOpt(body, "version", proc.version);
    setOpt(body, "status", proc.status);
    setOpt(body, "description", proc.description);
    setOpt(body, "confidence", proc.confidence);
    setOpt(body, "source", proc.source);
    setOpt(body, "constraints", proc.constraints);
    return this.http.post(`${this.ns}/upsert_procedure`, body);
  }

  // ── Query ────────────────────────────────────────────────────────────────

  /**
   * Query across episodic, semantic, and procedural memory layers. Mode is
   * auto-detected from the query text when not specified.
   */
  query(
    query: string,
    opts: MemoryQueryOptions = {},
  ): Promise<MemoryQueryResult> {
    const body: Record<string, unknown> = { query };
    setOpt(body, "entity_id", opts.entity_id);
    setOpt(body, "session_id", opts.session_id);
    setOpt(body, "procedure_id", opts.procedure_id);
    setOpt(body, "current_node_id", opts.current_node_id);
    setOpt(body, "context", opts.context);
    setOpt(body, "mode", opts.mode);
    setOpt(body, "limit", opts.limit);
    setOpt(body, "include_evidence", opts.include_evidence);
    setOpt(body, "include_plan", opts.include_plan);
    setOpt(body, "include_diagnostics", opts.include_diagnostics);
    return this.http.post(`${this.ns}/query`, body);
  }

  // ── Next step ────────────────────────────────────────────────────────────

  /**
   * Return the next valid node in a procedure given the current node and
   * evaluation context.
   */
  nextStep(
    procedureId: string,
    opts: NextStepOptions = {},
  ): Promise<ProcedureStep> {
    const body: Record<string, unknown> = { procedure_id: procedureId };
    setOpt(body, "current_node_id", opts.current_node_id);
    setOpt(body, "context", opts.context);
    return this.http.post(`${this.ns}/next_step`, body);
  }

  // ── Timeline ─────────────────────────────────────────────────────────────

  /** Fetch the episodic timeline for an entity, ordered chronologically. */
  timeline(entityId: string, limit?: number): Promise<MemoryRecord[]> {
    const params: Record<string, string | number | boolean | undefined> = {};
    if (limit !== undefined) params["limit"] = limit;
    return this.http.get(
      `${this.ns}/timeline/${encodeURIComponent(entityId)}`,
      params,
    );
  }

  // ── Graph edges ──────────────────────────────────────────────────────────

  /** Create a typed weighted edge between two memory records. */
  createEdge(req: CreateEdgeRequest): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {
      source_id: req.source_id,
      target_id: req.target_id,
      edge_type: req.edge_type,
    };
    setOpt(body, "id", req.id);
    setOpt(body, "weight", req.weight);
    setOpt(body, "metadata", req.metadata);
    return this.http.post(`${this.ns}/edges`, body);
  }

  /** Return all outgoing and incoming edges for a memory node. */
  nodeEdges(memoryId: string): Promise<NodeEdgesResult> {
    return this.http.get(
      `${this.ns}/edges/${encodeURIComponent(memoryId)}`,
    );
  }

  /** Delete a memory edge by its ID. */
  deleteEdge(edgeId: string): Promise<Record<string, unknown>> {
    return this.http.post(
      `${this.ns}/edges/${encodeURIComponent(edgeId)}/delete`,
    );
  }

  // ── Belief versioning ────────────────────────────────────────────────────

  /** Return the full version history of a semantic fact by its fact_key. */
  beliefHistory(factKey: string): Promise<BeliefVersion[]> {
    return this.http.get(
      `${this.ns}/beliefs/${encodeURIComponent(factKey)}/history`,
    );
  }

  // ── Centrality ───────────────────────────────────────────────────────────

  /** Recompute PageRank centrality scores for all nodes in the namespace. */
  refreshCentrality(): Promise<CentralityResult> {
    return this.http.post(`${this.ns}/graph/centrality`);
  }
}

function setOpt(
  body: Record<string, unknown>,
  key: string,
  value: unknown,
): void {
  if (value !== undefined && value !== null) {
    body[key] = value;
  }
}
